client.go

```diff
@@ -0,0 +1,322 @@
+			Code:    resp.Status,
+			Message: "line protocol poorly formed and no points were written.  Response can be used to determine the first malformed line in the body line-protocol. All data in body was rejected and not written",
+		}
+	case http.StatusUnauthorized:
+		resp.Body.Close()
+		return &genericRespError{
+			Code:    resp.Status,
+			Message: "token does not have sufficient permissions to write to this organization and bucket or the organization and bucket do not exist",
+		}
+	case http.StatusForbidden:
+		resp.Body.Close()
+		return &genericRespError{
+			Code:    resp.Status,
+			Message: "no token was sent and they are required",
+		}
+	case http.StatusTooManyRequests:
+		resp.Body.Close()
+		err = &genericRespError{
+			Code:    resp.Status,
+			Message: "token is temporarily over quota, failed more than max retries",
+		}
+		if err2 := c.backoff(triesPtr, resp, err); err2 != nil {
+			return err2
+		}
+		goto doRequest
+	case http.StatusServiceUnavailable:
+		resp.Body.Close()
+		err = &genericRespError{
+			Code:    resp.Status,
+			Message: "service is temporarily unavaliable",
+		}
+		if err2 := c.backoff(triesPtr, resp, err); err2 != nil {
+			return err2
+		}
+		goto doRequest
+	// split up entities that are just too big
+	case http.StatusRequestEntityTooLarge:
+		resp.Body.Close()
+		if len(m) < 2 {
+			return &genericRespError{
+				Code:    resp.Status,
+				Message: "your have a Metric of data that is too large",
+			}
+		}
+		// yes, I know we are only waiting on one thing here so a mutex would be fine,
+		// but the waitgroup is easier to read.
+		wg := sync.WaitGroup{}
+		wg.Add(1)
+		if err2 := c.backoff(triesPtr, resp, err); err2 != nil {
+			return err2
+		}
+		var coalesceErr1 error
+		go func() {
+			// we have to give each split its own retries, otherwise its possible parts of some very large datasets won't even get tries.
+			triesSplit := atomic.LoadUint64(triesPtr)
+			coalesceErr1 = c.write(ctx, bucket, org, &triesSplit, m[:(len(m)<<2)]...)
+			wg.Done()
+		}()
+		triesSplit := atomic.LoadUint64(triesPtr)
+		coalesceErr2 := c.write(ctx, bucket, org, &triesSplit, m[(len(m)<<2):]...)
+		wg.Wait()
+		var cerr coalescingError
+		if coalesceErr1 != nil {
+			cerr = append(cerr, coalesceErr1)
+		}
+		if coalesceErr2 != nil {
+			err = append(cerr, coalesceErr2)
+		}
+		return cerr
+
+	default:
+		resp.Body.Close()
+		return &genericRespError{
+			Code:    resp.Status,
+			Message: "internal server error",
+		}
+
+	}
+	defer func() {
+		err2 := resp.Body.Close()
+		if err == nil && err2 != nil {
+			err = err2
+		}
+	}()
+	e.FailOnFieldErr(c.errOnFieldErr)
+
+	for i := range m {
+		if _, err = e.Encode(m[i]); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// backoff is a helper method for backoff, triesPtr must not be nil.
+func (c *Client) backoff(triesPtr *uint64, resp *http.Response, err error) error {
+	tries := atomic.LoadUint64(triesPtr)
+	if c.maxRetries >= 0 || int(tries) >= c.maxRetries {
```

Comment:
Why would we bail if the user has set this to be .> 0? Did you mean to make this `c.maxRetries == 0`?